rllib/core/models/tf/encoder.py

```diff
@@ -7,54 +7,124 @@
-from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.framework import try_import_tf
 from ray.rllib.utils.nested_dict import NestedDict
 
-torch, nn = try_import_torch()
+_, tf, _ = try_import_tf()
 
 
 class TfMLPEncoder(Encoder, TfModel):
-    def __init__(self, config: ModelConfig) -> None:
+    def __init__(self, config: MLPEncoderConfig) -> None:
         TfModel.__init__(self, config)
         Encoder.__init__(self, config)
 
-        # Create the neural networks
+        # Create the neural network.
         self.net = TfMLP(
             input_dim=config.input_dims[0],
             hidden_layer_dims=config.hidden_layer_dims,
-            output_dim=config.output_dims[0],
             hidden_layer_activation=config.hidden_layer_activation,
+            hidden_layer_use_layernorm=config.hidden_layer_use_layernorm,
+            output_dim=config.output_dims[0],
+            output_activation=config.output_activation,
+            use_bias=config.use_bias,
+        )
+
+    @override(Model)
+    def get_input_specs(self) -> Union[Spec, None]:
+        return SpecDict(
+            {
+                SampleBatch.OBS: TfTensorSpec("b, d", d=self.config.input_dims[0]),
+                STATE_IN: None,
+                SampleBatch.SEQ_LENS: None,
+            }
+        )
+
+    @override(Model)
+    def get_output_specs(self) -> Union[Spec, None]:
+        return SpecDict(
+            {
+                ENCODER_OUT: TfTensorSpec("b, d", d=self.config.output_dims[0]),
+                STATE_OUT: None,
+            }
         )
 
     @override(Model)
-    def get_input_spec(self) -> Union[Spec, None]:
+    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+        return NestedDict(
+            {
+                ENCODER_OUT: self.net(inputs[SampleBatch.OBS]),
+                STATE_OUT: inputs[STATE_IN],
+            }
+        )
+
+
+class TfCNNEncoder(TfModel, Encoder):
+    def __init__(self, config: CNNEncoderConfig) -> None:
+        TfModel.__init__(self, config)
+        Encoder.__init__(self, config)
+
+        layers = []
+        # The bare-bones CNN (no flatten, no succeeding dense).
+        cnn = TfCNN(
+            input_dims=config.input_dims,
+            cnn_filter_specifiers=config.cnn_filter_specifiers,
+            cnn_activation=config.cnn_activation,
+            cnn_use_layernorm=config.cnn_use_layernorm,
+            use_bias=config.use_bias,
+        )
+        layers.append(cnn)
+        # Flatten output of last CNN layer.
+        layers.append(tf.keras.layers.Flatten())
+
+        # Add a final linear layer to make sure that the outputs have the correct
+        # dimensionality (output_dims).
+        output_activation = get_activation_fn(config.output_activation, framework="tf2")
+        layers.append(
+            tf.keras.layers.Dense(config.output_dims[0], activation=output_activation),
+        )
+
+        self.net = tf.keras.Sequential(layers)
```

Comment:
Nit: Can we unify the constructor a little more? So that the order of things and the comments are the same between Tf and Torch?

Reply:
@kouroshHakha I've been thinking about our specs and think we should pull the functionality of TfTensorSpec and TorchTensorSpec into TensorSpec and give TensorSpec a framework kwarg.

If the kwarg is None, simply don't enforce tensor framework and check based on the incoming tensor.
That way we could unify our specs..
-> Saves many LOCs
-> Reduce burden auf maintanance and always checking for equality between frameworks

Reply:
Especially when rolling this out over RLLib and writing many new RLModules and possibly Models over time, the sepcs would become less of a burden.
Wdyt?

Reply:
This would get us another step close to having 99% of RLModule and Model code in base clases and having the framework specific classes only be separated by an attribute self.framework.
We are almost there for the PPORLModule.

Reply:
Love the idea. Let's do this!

Reply:
unified c'tors a little more, comments, structure, etc..